src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,936 @@
+	 * variable is given
+	 * cascateToColocated determines whether to cascated
+	 * shardCount and colocateWith will be cascaded to the
+	 * currently colocated tables
+	 */
+	bool cascadeToColocatedIsNull;
+	bool cascadeToColocated;
+
+	/* schema of the table */
+	char *schemaName;
+	Oid schemaId;
+
+	/* name of the table */
+	char *relationName;
+
+	/* new relation oid after the conversion */
+	Oid newRelationId;
+
+	/* temporary name for intermediate table */
+	char *tempName;
+
+	/*hash that is appended to the name to create tempName */
+	uint32 hashOfName;
+
+	/* shard count of the table before conversion */
+	int originalShardCount;
+
+	/* list of the table oids of tables colocated with the table before conversion */
+	List *colocatedTableList;
+
+	/* new distribution key, if distributionColumn variable is given */
+	Var *distributionKey;
+
+	/* distribution key of the table before conversion */
+	Var *originalDistributionKey;
+
+	/*
+	 * The function that will be used for the conversion
+	 * Must comply with conversionType
+	 * UNDISTRIBUTE_TABLE -> UndistributeTable
+	 * ALTER_DISTRIBUTED_TABLE -> AlterDistributedTable
+	 * ALTER_TABLE_SET_ACCESS_METHOD -> AlterTableSetAccessMethod
+	 */
+	TableConversionFunction function;
+} TableConversion;
+
+
+static void UndistributeTable(TableConversion *con);
+static void AlterDistributedTable(TableConversion *con);
+static void AlterTableSetAccessMethod(TableConversion *con);
+static void ConvertTable(TableConversion *con);
+static void EnsureTableNotReferencing(Oid relationId);
+static void EnsureTableNotReferenced(Oid relationId);
+static void EnsureTableNotForeign(Oid relationId);
+static void EnsureTableNotPartition(Oid relationId);
+static TableConversion * CreateTableConversion(char conversionType, Oid relationId,
+											   char *distributionColumn,
+											   bool shardCountIsNull, int shardCount,
+											   char *colocateWith, char *accessMethod,
+											   bool cascadeToColocatedIsNull,
+											   bool cascadeToColocated);
+static void CreateDistributedTableLike(TableConversion *con);
+static void CreateCitusTableLike(TableConversion *con);
+static List * GetViewCreationCommandsOfTable(Oid relationId);
+static void ReplaceTable(Oid sourceId, Oid targetId, List *justBeforeDropCommands);
+static void AlterDistributedTableMessages(TableConversion *con);
+
+PG_FUNCTION_INFO_V1(undistribute_table);
+PG_FUNCTION_INFO_V1(alter_distributed_table);
+PG_FUNCTION_INFO_V1(alter_table_set_access_method);
+
+
+/*
+ * undistribute_table gets a distributed table name and
+ * udistributes it.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+	EnsureRelationExists(relationId);
+	EnsureTableOwner(relationId);
+
+	TableConversion *con = CreateTableConversion(UNDISTRIBUTE_TABLE, relationId, NULL,
+												 true, 0, NULL, NULL, true, false);
```

Comment:
the TableConversion struct seems to be a mixture of input parameters and state.

It might be nicer to separate those, e.g. construct a TableConversionParameters here, pass it to UndistributeTable, and then UndistributeTable creates a TableConversion (or to be 
very explicit: TableConversionState). The TableConversionState probably contains the original TableConversionParameters. 

If the TableConversionParameters is a lot cleaner, you can also consider inlining it, e.g.

```c
TableConversionParameters conversionParameters = {
    .conversionType = UNDISTRIBUTE_TABLE,
    .relationId = relationId,
};

UndistributeTable(&conversionParameters);
```